### PR TITLE
Rocq CLI prototype (exec based)

### PR DESCRIPTION
**Pull request by SkySkimmer**
Executable `rocq` can be called as `rocq compile args` (also `rocq c
args`) and will in turn call `coqworker --kind=compile
args`. (coqworker will be renamed later)

coqc has also been modified to call coqworker.

The design where `rocq` calls the worker executable instead of being
itself the worker executable is so that `rocq check` can work without
involving the main worker code.

We could also consider not having `rocq` handle `rocqchk` ie keeping
it completely separate.

TODO:
- [x] fix makefile generated by rocq makefile to call `rocq foo` instead of `coqfoo`
- [x] overlays 
-  [x] update test suite to use rocq

Issues to be solved:

- ~dune doesn't understand the calling `rocq c` involves
  `coqworker` so if we start using it in our generated rules for
  stdlib the dependencies will be incorrect.
  Similarly coqc implicitly calling the worker breaks dune's coq
  mode's dependencies.~

  solution: make coqdep emit a dependency `foo.vo: coqworker`

Depends: 
- https://github.com/coq/coq/pull/19763

Overlays:
- https://github.com/ejgallego/coq-lsp/pull/883
- https://github.com/coq-community/coq-dpdgraph/pull/138

### CODE REVIEW

**Comment by SkySkimmer** [comment]
@coqbot run full ci

File: topbin/dune [hunk 1]
```diff
@@ -18,7 +24,10 @@
  (public_name coqc)
  (package coq-core)
  (modules coqc_bin)
- (libraries coq-core.toplevel))
+ ; we include the hash of coqworker.byte but actually depend on coqworker.opt
+ (link_deps %{project_root}/_build/install/%{context_name}/lib/coq-core/coqworker.opt%{ext_exe})
```
Comment (SkySkimmer) [inline topbin/dune:28]: It seems this dependency breaks dune when the coq rules for stdlib are present. Is there some workaround?

@ejgallego @rgrinberg 

(context: `coqc` calls `coqworker` at runtime in this PR)

**Comment by SkySkimmer** [comment]
instead of putting the hash of coqworker in coqc it may make more sense to have coqdep output a dep `foo.vo : coqworker`

this appears to work except it breaks some of our tests as they diff the output of coqdep

@coqbot run full ci

**Comment by SkySkimmer** [comment]
macos failure is https://github.com/coq/coq/issues/19862 (we do `COQLIB="$(pwd)/../_install_ci/lib/coq/"` and `$(pwd)` has a trailing slash producing the bugged `//..`)

**Comment by SkySkimmer** [comment]
I don't understand what's going on in windows, it seems like the `execv` does nothing? The log says
~~~
(cd _build/default && C:\ci\cygwin64\opam\CP.2023.11.0~dev\bin\coqc.exe -noinit -boot -R theories Stdlib -w +default -q -w -deprecated-native-compiler-option -native-compiler off -time-file theories\Init\Notations.timing theories/Init/Notations.v)
OCAMLPATH = C:\ci\cygwin64\opam\CP.2023.11.0~dev\.opam-switch\build\coq-stdlib.dev\_build/install/default/lib
Using C:\ci\cygwin64\opam\CP.2023.11.0~dev\lib\coq-core\coqworker.opt.exe
File "theories/dune", lines 858-861, characters 1-364:
858 |  (rule
859 |   (targets Notations.timing Notations.glob Notations.vos Notations.vo)
860 |   (deps C:/ci/cygwin64/opam/CP.2023.11.0~dev/lib/coq-core/coqworker.opt.exe)
861 |   (action (chdir ..\..\. (run coqc -noinit -boot -R theories Stdlib -w +default -q -w -deprecated-native-compiler-option -native-compiler off -time-file theories\Init\Notations.timing %{dep:Notations.v}))))
Error: Rule failed to generate the following targets:
- theories/Init/Notations.glob
- theories/Init/Notations.timing
- theories/Init/Notations.vo
- theories/Init/Notations.vos
~~~
if the execv in coqc_bin was working there should be a `HELLO WORLD FROM WORKER` after the `Using C:\ci\cygwin64\opam\CP.2023.11.0~dev\lib\coq-core\coqworker.opt.exe`.

Is Unix.execv broken on windows?

cc @MSoegtropIMC

**Comment by MSoegtropIMC** [comment]
Sorry - the log looks so contrived that I only saw on second glance that it is plain Coq Platform build environment.

Why are there so many warnings in there?

**Comment by MSoegtropIMC** [comment]
I made a local run.

Configuration:
```
SET GITHUB_SERVER_URL=https://github.com
SET GITHUB_REPOSITORY=SkySkimmer/coq
SET GITHUB_SHA=5c159e1ac944738bd3d56baa6489a6755e7a71ca

CALL coq_platform_make_windows.bat ^
  -arch=64 ^
  -destcyg=E:\bin\cygwin64_coq_platform_gaetan ^
  -cygcache=C:\bin\cygwin_cache ^
  -cygrepo=https://mirrors.kernel.org/sourceware/cygwin ^
  -packages="8.19~2024.10" -extent=b -parallel=p -jobs=8 -switch=k ^
  -override-dev-pkg="coq-core=%GITHUB_SERVER_URL%/%GITHUB_REPOSITORY%/archive/%GITHUB_SHA%.tar.gz" ^
  -override-dev-pkg="coq-stdlib=%GITHUB_SERVER_URL%/%GITHUB_REPOSITORY%/archive/%GITHUB_SHA%.tar.gz" ^
  -override-dev-pkg="coq=%GITHUB_SERVER_URL%/%GITHUB_REPOSITORY%/archive/%GITHUB_SHA%.tar.gz" ^
  -override-dev-pkg="coqide-server=%GITHUB_SERVER_URL%/%GITHUB_REPOSITORY%/archive/%GITHUB_SHA%.tar.gz" ^
  -override-dev-pkg="coqide=%GITHUB_SERVER_URL%/%GITHUB_REPOSITORY%/archive/%GITHUB_SHA%.tar.gz" ^
```

Observations:
- I do see `HELLO WORLD FROM WORKER`
- I did run the failing dune command in the opam switch build folder with `--verbose` to get the failing coqc command. The output of this is:
```
/opam/CP.2024.10.0~8.19~2024.10/.opam-switch/build/coq-stdlib.dev
$ coqc.exe -noinit -boot -R theories Stdlib -w +default -q -w -deprecated-native-compiler-option -native-compiler off -time-file theories\Init\Logic.timing theories/Init/Logic.v
OCAMLPATH =
Using E:\bin\cygwin64_coq_platform_gaetan\opam\CP.2024.10.0~8.19~2024.10\lib\coq-core\coqworker.opt.exe
HELLO WORLD FROM WORKER
File ".\theories/Init/Logic.v", line 13, characters 15-24:
Error: Cannot find a physical path bound to logical path Notations.
```

**Comment by SkySkimmer** [comment]
did you run that command in `_build/default` ? otherwise that's normal
in the generated theories/dune does the rule for Logic.vo have the right dependencies? it should look like
~~~
(rule
  (targets Logic.timing Logic.glob Logic.vos Logic.vo)
  (deps %{project_root}/_build/install/default/lib/coq-core/coqworker.opt
        ../../theories/Init/Notations.vo ../../theories/Init/Ltac.vo)
  (action (chdir ../../. (run coqc -noinit -boot -R theories Stdlib -w +default -q -w -deprecated-native-compiler-option -native-compiler off -time-file theories/Init/Logic.timing %{dep:Logic.v}))))
~~~
(maybe with some \ instead of /)

**Comment by MSoegtropIMC** [comment]
Yes, sorry, I was too quick. When I run it in the correct folder it actually succeeds and also creates the .vo file:
```
Michael@MININT /opam/CP.2024.10.0~8.19~2024.10/.opam-switch/build/coq-stdlib.dev/_build/default
$ rm theories/Init/Logic.vo

Michael@MININT /opam/CP.2024.10.0~8.19~2024.10/.opam-switch/build/coq-stdlib.dev/_build/default
$ coqc.exe -noinit -boot -R theories Stdlib -w +default -q -w -deprecated-native-compiler-option -native-compiler off -time-file theories\Init\Logic.timing theories/Init/Logic.v
OCAMLPATH =
Using E:\bin\cygwin64_coq_platform_gaetan\opam\CP.2024.10.0~8.19~2024.10\lib\coq-core\coqworker.opt.exe
HELLO WORLD FROM WORKER

Michael@MININT /opam/CP.2024.10.0~8.19~2024.10/.opam-switch/build/coq-stdlib.dev/_build/default
$ echo $?
0

Michael@MININT /opam/CP.2024.10.0~8.19~2024.10/.opam-switch/build/coq-stdlib.dev/_build/default
$ ls -l theories/Init/Logic.vo
-rwxr-xr-x 1 Michael None 128050 22. Nov 18:45 theories/Init/Logic.vo

Michael@MININT /opam/CP.2024.10.0~8.19~2024.10/.opam-switch/build/coq-stdlib.dev/_build/default
```

**Comment by MSoegtropIMC** [comment]
But if I run dune I get:
```
Michael@MININT /opam/CP.2024.10.0~8.19~2024.10/.opam-switch/build/coq-stdlib.dev
$ dune.exe build --verbose -p coq-stdlib -j 19 --promote-install-files=false @install
Shared cache: disabled
Shared cache location:
C:\Users\Michael\AppData\Local\Microsoft\Windows\INetCache\dune\db
Workspace root:
"E:\\bin\\cygwin64_coq_platform_gaetan\\opam\\CP.2024.10.0~8.19~2024.10\\.opam-switch\\build\\coq-stdlib.dev"
Dune context:
 { name = "default"
 ; kind = "default"
 ; profile = Release
 ; merlin = true
 ; fdo_target_exe = None
 ; build_dir = In_build_dir "default"
 ; instrument_with = []
 }
Actual targets:
- recursive alias @install
Running[1]: (cd _build/default && E:\bin\cygwin64_coq_platform_gaetan\opam\CP.2024.10.0~8.19~2024.10\bin\coqc.exe -noinit -boot -R theories Stdlib -w +default -q -w -deprecated-native-compiler-option -native-compiler off -time-file theories\Init\Logic.timing theories/Init/Logic.v)
Output[1]:
OCAMLPATH = E:\bin\cygwin64_coq_platform_gaetan\opam\CP.2024.10.0~8.19~2024.10\.opam-switch\build\coq-stdlib.dev\_build/install/default/lib
Using E:\bin\cygwin64_coq_platform_gaetan\opam\CP.2024.10.0~8.19~2024.10\lib\coq-core\coqworker.opt.exe
File "theories/dune", lines 840-844, characters 1-440:
840 |  (rule
841 |   (targets Logic.timing Logic.glob Logic.vos Logic.vo)
842 |   (deps E:/bin/cygwin64_coq_platform_gaetan/opam/CP.2024.10.0~8.19~2024.10/lib/coq-core/coqworker.opt.exe
843 |         ..\..\theories\Init\Notations.vo ..\..\theories\Init\Ltac.vo)
844 |   (action (chdir ..\..\. (run coqc -noinit -boot -R theories Stdlib -w +default -q -w -deprecated-native-compiler-option -native-compiler off -time-file theories\Init\Logic.timing %{dep:Logic.v}))))
Error: Rule failed to generate the following targets:
- theories/Init/Logic.glob
- theories/Init/Logic.timing
- theories/Init/Logic.vo
- theories/Init/Logic.vos
```

**Comment by MSoegtropIMC** [comment]
But I have:
```
Michael@MININT /opam/CP.2024.10.0~8.19~2024.10/.opam-switch/build/coq-stdlib.dev
$ ls -l _build/default/theories/Init/Logic.*
-rwxr-xr-x 1 Michael None 191081 22. Nov 18:48 _build/default/theories/Init/Logic.glob
-rwxr-xr-x 1 Michael None  39058 22. Nov 18:48 _build/default/theories/Init/Logic.timing
-r-xr-xr-x 1 Michael None  41561 22. Nov 18:27 _build/default/theories/Init/Logic.v
-rwxr-xr-x 1 Michael None 128050 22. Nov 18:48 _build/default/theories/Init/Logic.vo
-rwxr-xr-x 1 Michael None      0 22. Nov 18:48 _build/default/theories/Init/Logic.vok
-rwxr-xr-x 1 Michael None      0 22. Nov 18:48 _build/default/theories/Init/Logic.vos
```

**Comment by MSoegtropIMC** [comment]
So all files dune complains about not being there are there. It looks like a bug in dune.

**Comment by MSoegtropIMC** [comment]
I will see what good old ProcMon gives some insight into what dune is up to.

**Comment by SkySkimmer** [comment]
I guess on windows Unix.execv (which is _wexecv https://github.com/ocaml/ocaml/blob/c484c6932fa2eae03ba0f5a7dbdb26e3eee65eb0/runtime/caml/misc.h#L508 https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/execv-wexecv) creates a new process and quits the current one, and dune doesn't wait for the new process?

**Comment by MSoegtropIMC** [comment]
Hmm OK, but why does it work since 10 years then? What is the difference between your build and Coq master?

**Comment by SkySkimmer** [comment]
Coq master: coqc does some work
this PR: coqc does exec coqworker, coqworker does some work

**Comment by MSoegtropIMC** [comment]
According to the Procmon logs dune does not try to access the Logic.vo file before the coq worker even started. dune looks at them, the coq worker looks at them and that was it then. Try to get a CSV dump, but the CSV file I exported is unreadable for whatever reason.

**Comment by MSoegtropIMC** [comment]
Yes, this might be it then.

**Comment by MSoegtropIMC** [comment]
Screenshot 2024-11-22 190551

**Comment by MSoegtropIMC** [comment]
Sorry I can only get a screen shot right now. This shows all access to files with "Logic.vo" in the name during the complete call of dune.

**Comment by MSoegtropIMC** [comment]
So it is as you say - dune looks for `Logic.vo` before coqworker writes the .vo file. I now extended the filter to include `Logic.v`. This shows that coqworker starts shortly before dune looks for the `.vo` file, but writes the `.vo` file quite a bit later.

**Comment by SkySkimmer** [comment]
@coqbot bench

**Comment by SkySkimmer** [comment]
@coqbot bench

**Comment by SkySkimmer** [comment]
@coqbot bench

**Comment by SkySkimmer** [comment]
Continued in https://github.com/coq/coq/pull/19927

**Pull request closed without merging**